extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/transaction/interceptor/ReactiveMongoTransaction.java

```diff
@@ -0,0 +1,98 @@
+package io.quarkus.mongodb.panache.transaction.interceptor;
+
+import javax.transaction.HeuristicMixedException;
+import javax.transaction.HeuristicRollbackException;
+import javax.transaction.RollbackException;
+import javax.transaction.Status;
+import javax.transaction.Synchronization;
+import javax.transaction.SystemException;
+import javax.transaction.Transaction;
+import javax.transaction.xa.XAResource;
+
+import com.mongodb.reactivestreams.client.ClientSession;
+
+public class ReactiveMongoTransaction implements Transaction {
+    private ClientSession clientSession;
+    private int status = Status.STATUS_ACTIVE;
+
+    public ClientSession getClientSession() {
+        return clientSession;
+    }
+
+    void setClientSession(ClientSession clientSession) {
+        this.clientSession = clientSession;
+    }
+
+    @Override
+    public void commit() throws RollbackException, HeuristicMixedException, HeuristicRollbackException, SecurityException,
+            IllegalStateException, SystemException {
+        if (isDone()) {
+            throw new IllegalStateException("Transaction is done. Cannot commit transaction.");
+        }
+
+        this.status = Status.STATUS_COMMITTING;
+        try {
+            this.clientSession.commitTransaction();
```

Comment:
Note that this returns a `Publisher<Void>` (https://mongodb.github.io/mongo-java-driver/4.1/apidocs/mongodb-driver-reactivestreams/com/mongodb/reactivestreams/client/ClientSession.html#commitTransaction()) and so it turns out that we can't implement this with the JTA transaction API.

I didn't realise this until today, so let's continue the discussion on quarkus-dev as soon as you start the thread. It's likely we'll get rid of JTA for reactive transactions and implement it differently, but introducing an abstraction for both Hibernate Reactive and Mongodb Reactive (if we can).

Reply:
Thanks, you just point me to why my entities are not persisted when using transaction, it's because the `commitTransaction` is async !

Reply:
:)

Reply:
I wait on the `commitTransacton` and `rollbackTransaction` for now on and this work fine.

There is now three places where I block, I add a comment `//FIXME we block here` as I think it's wrong to block on an IO inside the interceptor for reactive MongoDB.
I'll write someting to the ML tomorrow to discuss this.